src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -3592,6 +3590,46 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         return nullptr;
     }
 
+    ASR::do_loop_head_t make_do_loop_head(ASR::expr_t *loop_start, ASR::expr_t *loop_end,
+                                            ASR::expr_t *inc, int a_kind, const Location& loc) {
+        ASR::do_loop_head_t head;
+        ASR::ttype_t *a_type = ASRUtils::TYPE(ASR::make_Integer_t(al, loc,
+            a_kind, nullptr, 0));
+        if (loop_start) {
+            head.m_start = loop_start;
+        } else {
+            head.m_start = ASR::down_cast<ASR::expr_t>(ASR::make_IntegerConstant_t(al, loc, 0, a_type));
+        }
+        if (inc) {
+            head.m_increment = inc;
+        } else {
+            ASR::expr_t *constant_one = ASR::down_cast<ASR::expr_t>(ASR::make_IntegerConstant_t(
+                                                al, loc, 1, a_type));
+            head.m_increment = constant_one;
+        }
+        // Loop end depends upon the sign of m_increment.
+        // if inc > 0 then: loop_end -=1 else loop_end += 1
+
+        // Compute the sign value of m_increment
+        ASR::symbol_t *fn_sign = resolve_intrinsic_function(loc, "_find_sign");
```

Comment:
This approach will require us to replace all the `for` loops in `_lpython_builtin.py` to be replaced by `while` loops in order to avoid an infinite loop.

Reply:
@certik @czgdp1807 Should we move forward with this approach?

Reply:
I would like to see all the 216 cases tested first. It should be easy to do it if you use loops. I can't say for sure (without looking at the tests) whether any approach will work or not. But yes its making sense to me if I read your code. Branching can be avoided in `_find_sign`'s implementation?

Reply:
> This approach will require us to replace all the `for` loops in `_lpython_builtin.py` by `while` loops in order to avoid an infinite loop.

For testing, we will require this change. If the approach looks good, it will indeed run for all 216 cases.

Reply:
Yes. Please test it. It would be great to see this approach working.

Reply:
If compile time values are hard to test, you can also test for runtime values only which is easy. Just create 3 lists for start, stop and end and send each combination into a function which tests these three in a loop. If it works, we will know everything is correct for runtime values.